myia/frontends/pytorch.py

```diff
@@ -0,0 +1,345 @@
+from ..abstract.data import AbstractClassBase, AbstractArray, AbstractScalar, \
+    ANYTHING, VALUE, TYPE, SHAPE
+from ..dtype import Int, UInt, Float, Bool, Number
+from ..utils import overload
+from ..pipeline.resources import standard_object_map, standard_method_map
+from ..hypermap import hyper_map
+from ..api import _convert_arg_init
+from ..pipeline.steps import convert_arg, convert_result
+
+from . import Frontend
+from .pytorch_functions import linear, tensor_dim, t
+
+
+_type_map = {
+    torch.int8: Int[8],
+    torch.int16: Int[16],
+    torch.int32: Int[32],
+    torch.int64: Int[64],
+    torch.uint8: UInt[8],
+    torch.float16: Float[16],
+    torch.float32: Float[32],
+    torch.float64: Float[64],
+    torch.uint8: Bool,
+    # This is a hack but we really need uint64 support
+    # torch.int64: Int[64],
+}
+
+
+def pytorch_dtype_to_type(dtype):
+    """Map a pytorch dtype to a myia type."""
+    if dtype not in _type_map:
+        raise TypeError(f"Unsupported dtype {dtype}")
+    return _type_map[dtype]
+
+
+pytorch_object_map = standard_object_map.copy()
+pytorch_object_map.update({
+    torch.nn.functional.linear: linear,
+    torch.tanh: C.tanh,
+})
+
+
+class AbstractModule(AbstractClassBase):
+    """Represents a PyTorch Module."""
+
+    def user_defined_version(self):
+        """Return the user-defined version of this type.
+
+        This uses the attribute types as defined by the user, rather than what
+        is generated by the inferrer or other methods.
+
+        Current default is to return self in order to make it easier for Myia
+        hypermap mapping function to return a different type from its input
+        (especially for pytorch modules and their contents).
+        """
+        return self
+
+
+class AbstractTensor(AbstractArray):
+    """Represents a PyTorch Tensor."""
+
+    def __init__(self, element, values, requires_grad=None, retain_grad=None):
+        """Initialize an AbstractTensor."""
+        super().__init__(element, values)
+        self.requires_grad = requires_grad
+
+        # TODO: IS RETAIN_GRAD EVEN NEEDED? IDK YET.
+        self.retain_grad = retain_grad
+
+
+'''
+class AbstractParameter(AbstractTensor):
+    """Represents a PyTorch Parameter."""
+
+    # This might end up as an alternative way to log that this is a Parameter.
+'''
+
+
+class TensorWrapper(ArrayWrapper):
+    """Represents a PyTorch Parameter."""
+
+    def __init__(self, array, dtype, shape, backend,
+                 requires_grad=None, retain_grad=None):
+        """Initialize the TensorWrapper."""
+        super().__init__(array, dtype, shape, backend)
+        self.requires_grad = requires_grad
+
+        # TODO: IS RETAIN_GRAD EVEN NEEDED? IDK YET.
+        self.retain_grad = retain_grad
+
+
+AT = AbstractTensor(ANYTHING, {SHAPE: ANYTHING})
+
+
+@C.tanh.register(AT)
+@core
+def array_tanh(xs):
+    """Implementation of `array_tanh`."""
+    return P.array_map(P.scalar_tanh, xs)
```

Comment:
You should not have to do this, it should just use the registered `AbstractArray` implementation. If it doesn't, that's a bug in the current system.